cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/templates/etl/common/ETLTemplate.java

```diff
@@ -159,4 +162,110 @@ public void configureAdapter(String adapterName, T config, AdapterConfigurer con
       configurer.setResources(resources);
     }
   }
+
+  private void validateStages(String source, String sink, List<String> transforms) throws Exception {
+    ArrayList<Type> unresTypeList = Lists.newArrayListWithCapacity(transforms.size() + 2);
+    Type inType = Transform.class.getTypeParameters()[0];
+    Type outType = Transform.class.getTypeParameters()[1];
+
+    // Load the classes using the class names provided
+    sourceClass = Class.forName(source);
+    sinkClass = Class.forName(sink);
+    TypeToken sourceToken = TypeToken.of(sourceClass);
+    TypeToken sinkToken = TypeToken.of(sinkClass);
+
+    // Extract the source's output type
+    if (RealtimeSource.class.isAssignableFrom(sourceClass)) {
+      Type type = RealtimeSource.class.getTypeParameters()[0];
+      unresTypeList.add(sourceToken.resolveType(type).getType());
+    } else {
+      unresTypeList.add(sourceToken.resolveType(outType).getType());
+    }
+
+    // Extract the transforms' input and output type
+    for (String transform : transforms) {
+      Class<?> klass = Class.forName(transform);
+      transformClasses.add(klass);
+      TypeToken transformToken = TypeToken.of(klass);
+      unresTypeList.add(transformToken.resolveType(inType).getType());
+      unresTypeList.add(transformToken.resolveType(outType).getType());
+    }
+
+    // Extract the sink's input type
+    if (RealtimeSink.class.isAssignableFrom(sinkClass)) {
+      Type type = RealtimeSink.class.getTypeParameters()[0];
+      unresTypeList.add(sinkToken.resolveType(type).getType());
+    } else {
+      unresTypeList.add(sinkToken.resolveType(inType).getType());
+    }
+
+    // Invoke validation method with list of unresolved types
+    validateTypes(unresTypeList);
+  }
+
+  @VisibleForTesting
+  static void validateTypes(ArrayList<Type> unresTypeList) {
```

Comment:
can you add some comments with an example input and an example of the resolved list it builds?